RFC: Allow type inference for const or static (#3546)
rust-lang/rfcs: text/0000-const-type-inference.md

- Feature Name: const_type_inference
- Start Date: 2023-12-21
- RFC PR: [rust-lang/rfcs#3546](https://github.com/rust-lang/rfcs/pull/3546)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

Allow type inference for `const` or `static` when the type of the initial value is known.

# Motivation
[motivation]: #motivation

> Rust currently requires explicit type annotations for `const` and `static` items.

Reviewer: It would be good to write a little bit about why Rust is like this currently.

Reviewer 2: I agree, this should have a longer explanation of Rusts "rule" of "no inference in signatures", how this RFC is breaking it and why this is okay.

Reviewer 3: Great suggestion! I actually don't know why the design was made that way. Maybe someone from the Rust team can help explain?

The "type is missing in const" error was emitted from the parser so my guess would be that it was just difficult to infer types when consts were implemented.

Reviewer 4: It's like this currently because it was decided that all public API points should be "obviously semver stable" rather than "quick to type".

Reviewer 3: Thanks for explaining this. I've incorporated it into the RFC.

In simple cases, explicitly writing out
the type of the const seems trivial. However, this isn't always the case:

- Sometimes the constant's value is complex, making the explicit type overly verbose.
- In some cases, the type may be unnameable.
- When creating macros, the precise type might not be known to the macro author.
- Code generators may not have enough information to easily determine the type.

This change aims to make Rust code more concise and maintainable, especially in scenarios where the types of
const items are complicated or not easily expressible.

> 

Reviewer 5: It might also be helpful to add ergonomics to the motivation. This particularly impacts new Rust users coming from languages where most or all types are inferred.

Here is a draft paragraph:
```suggestion

Inferring constant types also improves the ergonomics of the language, particularly for new users. Some users are
coming from languages where most (or all) types are inferred. So inferring obvious types matches their
expectations. Other new users are focused on learning ownership, or other core Rust concepts. Reducing the
amount of boilerplate reduces their mental load. This reduction in mental load also helps experienced programmers.

```

Reviewer 3: Thank you!

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

You may declare constants and static variables without specifying their types when the type can be inferred
from the initial value. For example:

```rs

> const PI = 3.1415; // inferred as f64

Reviewer: I would personally prefer that the actual type of numeric or float types be known rather than picking arbitrary defaults (e.g. `i32` or `f64)`. I'm not too keen on it in a local context either tbh but there it's mitigated by the compiler being able to infer the real type from the surrounding code most of the time.
```rust
const PI = 3.1415; // error
const PI = 3.1415_f64; // inferred as f64
```

Reviewer 3: I would prefer more consistent behavior with let bindings.  So let's make a simple vote here:
🎉:  `const PI = 3.1415; // error`
🚀: `const PI = 3.1415; // inferred as f64`

Reviewer: It's not a big issue for me either way it's just that a few times I've only later realised a type has been unexpectedly made  `i32`. But it's easily fixed and more of an annoyance than a problem per se. It doesn't help that an `i32` is very rarely what I want.

Reviewer 4: If it follows the normal rust inference default types then it's at least no *worse* than what `let` does.

Reviewer: Well it's worse in the sense that let can use future code to infer the type, so mostly this is a non issue there unless there's a lot of generics involved.

Reviewer 4: Honestly I would most like if the const was of `{float}` type but people aren't ready for that conversation maybe

Reviewer 6: [Go `const`](https://go.dev/blog/constants)? :thinking:

static MESSAGE = "Hello, World!"; // inferred as &'static str

> const FN_PTR = std::string::String::default; // inferred as fn() -> String

Reviewer 7: technically the `const` would have the function item's type instead of a function pointer type.
https://doc.rust-lang.org/reference/types/function-item.html

Reviewer 3: This is actually a good point - I guess there's not much point to coerce the function item's type into function pointer type for `const` items. But should we coerce for `static` items?

Reviewer 4: I'm confused about why a person would want it as a static at all, so perhaps we shouldn't allow it at all in the first version of this.

Reviewer 8: @Lokathor maybe they just use the const as a quick switch between two cfg implementations?

Reviewer 4: It makes sense as a `const` but little sense as a `static`

Reviewer 7: when the type is a ZST, like function item types, at runtime, `const` and `static` are essentially interchangable. when the type is a function pointer, unless wrapped in something with interior mutability, `static` is basically just a `const` with a stable address, IIRC LLVM will still constant-propagate the value to most uses, since it's marked read-only so LLVM knows the value won't change.

Reviewer 2: Doing implicit coercions into an unknown target type sounds very confusing. If the right-hand side is `String::default`, then the resulting type should be the actual type of the expression, the function item ZST.

Today, the compiler already gives hint for most cases where the const or static item is missing a type:

```
802 | const A = 0;
    |        ^ help: provide a type for the constant: `: i32`
```

```
error: missing type for `const` item                                                     
  --> file.rs:27:26
   |
27 |     pub const update_blas = SystemStage { system: test_system, stage: vk::Pipeli... 
   |                          ^ help: provide a type for the constant: `: render_pass::SystemStage<for<'a> fn(ResMut<'a, AsyncQueues>)>`
```

The implementation should only need to carry over this information and set the type correspondingly
instead of emitting an error.

# Drawbacks
[drawbacks]: #drawbacks

- Potential Loss of Clarity: In some cases, omitting the type might make the code less clear,
  especially to newcomers or in codebases where explicit types are part of the documentation style.

>   It is my belief that this is a choice better left for the developers as in the case of `let` bindings.

Reviewer 5: ```suggestion
  It is my belief that this is a choice better left for the developers as in the case of `let` bindings.
  This drawback could be addressed using an allow-by-default clippy lint for `const` and `static` types.
```

- Anything else?

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

- Impact of Not Doing This: Rust code remains more verbose than necessary, especially in complex scenarios, and macro authors face challenges with type specifications.
- Alternative: Allowing the naming of function types as in [#3476](https://github.com/rust-lang/rfcs/pull/3476) may help resolve some of the cases where type inference is needed.

# Prior art
[prior-art]: #prior-art

In [RFC#1623](https://github.com/rust-lang/rfcs/pull/1623) we added `'static` lifetimes to every reference or generics lifetime value in `static` or `const` declarations.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

Should we allow assignment of unnameable types? For example,
```rs

> let A = |a: u32| {

Reviewer 2: ```suggestion
const A = |a: u32| {
```

Today, the compiler already gives hint for most cases where the const or static item is missing a type:

```
802 | const A = 0;
    |        ^ help: provide a type for the constant: `: i32`
```

```
error: missing type for `const` item                                                     
  --> file.rs:27:26
   |
27 |     pub const update_blas = SystemStage { system: test_system, stage: vk::Pipeli... 
   |                          ^ help: provide a type for the constant: `: render_pass::SystemStage<for<'a> fn(ResMut<'a, AsyncQueues>)>`
```

The implementation should only need to carry over this information and set the type correspondingly
instead of emitting an error.

# Drawbacks
[drawbacks]: #drawbacks

- Potential Loss of Clarity: In some cases, omitting the type might make the code less clear,
  especially to newcomers or when explicit types are needed to understanding the purpose of the item.
  It is my belief that this is a choice better left for the developers as in the case of `let` bindings.

> - Semvar compatibilty: It's a good idea that public API endpoints should be "obviously semvar stable".

Reviewer 2: ```suggestion
- Semver compatibilty: The API surface of the type is implicit, changing the right-hand side in subtle ways can change the type in a way that can be hard to notice, for example between different integer types. 
```

>   However, not all `const` or `static` items are public, and explicit typing isn't always important for semvar stability.

Reviewer 2: ```suggestion
  However, not all `const` or `static` items are public, and in many cases the type is obvious enough that semver isn't a concern.
```

  Requiring explicit typing for this reason seems a bit heavy handed.

> Both of these drawback could be addressed using an allow-by-default clippy lint for `const` and `static` types.

Reviewer 2: I don't think an allow-by-default lint addresses concerns about it being potentially confusing. Having a lint can be useful, but it doesn't address the problem because almost everyone won't be using it (but may be using inference).

Reviewer 5: Fair enough, happy to withdraw this suggestion and go with what you've suggested instead.

[guide-level-explanation]: #guide-level-explanation

You may declare constants and static variables without specifying their types when the type can be inferred
from the initial value, subjecting to the following constraints:

> - All numerical types contributing to the inference must be specified.

Reviewer 9: This description doesn't match the examples below. My reading of this would be that 

```rust
const PI: _ = 3.1415_f32
```
because it has an inferred type that is numeric. And as written, this rule feels wrong to me, because it treats numeric types specially.

How about something like:

 "the types of all literals must be fully constrained, which generally means numeric literals must either have a type suffix, or the type must specify their type"
 
 or
 
 "the type of any integer or floating point literal in the expression must have a type that is uniquely determine from the context"
 
 to mirror the wording from https://doc.rust-lang.org/reference/expressions/literal-expr.html
